README.md
# CLI

PHP Framework for command line interfaces

[![Latest Stable Version](https://img.shields.io/packagist/v/bheisig/cli.svg)](https://packagist.org/packages/bheisig/cli)
[![Minimum PHP Version](https://img.shields.io/badge/php-%3E%3D%207.0-8892BF.svg)](https://php.net/)
[![Build Status](https://travis-ci.org/bheisig/cli.svg?branch=master)](https://travis-ci.org/bheisig/cli)

##  About

You want an easy way to build a command line tool written in PHP? You are bored of big, over-engineered frameworks? And last but not least, you do not want to reinvent the wheel? -- Great, this could be the (almost) perfect solution for your next project.

The goal is to build intuitive CLI tools. Nobody likes to read documentation, so make your application self-descriptive!

##  Features

-   Easy-to-install via [Composer](https://getcomposer.org/)
-   Perfectly suited to create a single binary file from your application
-   Simple interface
-   Separate your features by commands
-   Runs on POSIX-like systems (GNU/Linux, MacOS, …) and Windows
-   Error/exception handling
-   Optional colored output based on log level
-   Log to standard or error output
-   Pre-defined usage output
-   JSON-based configuration with defaults, system-wide, user defined and runtime settings
-   Support for long and short options with optional or required values
-   Pre-defined options for verbosity/quietness, usage, version information, additional configuration files, runtime settings, configuration test
-   Verbosity mode outputs human-readable runtime in seconds and peak memory usage

## Example

For a simple application look at the `example/` folder.

## Requirements

-   PHP, version 7.1 or higher (7.0 works but is deprecated, 7.3 is recommended, unreleased 7.4 should work)
-   PHP modules `calendar`, `cli`, `date`, `filter` and `json`

## Installation

Go to your project folder and require this framework via Composer:

~~~ {.bash}
composer require bheisig/cli
~~~

## Update

Composer is the way to go:

~~~ {.bash}
composer update
~~~

## Copyright & License

Copyright (C) 2018-19 [Benjamin Heisig](https://benjamin.heisig.name/)

Licensed under the [GNU Affero GPL version 3 or later (AGPLv3+)](https://gnu.org/licenses/agpl.html). This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased][]

### Added

-   `HashMap`: Add methods to manipulate arrays
-   `Value`: Add methods to handle strings
-   `Service\Service`: Add base service with access to configuration and logging
-   `Service\UserInteraction`: Add service to handle user interaction
-   `Service\Shell`: Add service to execute shell commands
-   `Service\YAML`: Add service to decode/encode YAML

### Changed

-   Drop PHP 5.6 support; minimum required version is 7.0
-   Declare strict types
-   `Command\Executes::setup()`: Switch return statement from self to void
-   `Command\Executes::execute()`: Switch return statement from self to void
-   `Command\Executes::tearDown()`: Switch return statement from self to void

### Fixed

-   `Log`: Remove unwanted format tags from non-colored output
-   `App::invokeLogging()`: Fix default log level from invalid value `-1` to `31` (everything but debug)
-   `App::parseArguments()`: Do not miss arguments in combination with options
-   `Command\Command::tearDown()`: Print last debug message (spent time, consumed memory etc.) to STDERR
-   `composer`: Clarify which PHP extensions are required and which ones are suggested

## [0.3][] – 2019-05-08

### Added

-   Provide standardized exit codes to close application
-   Print stack trace in debug mode after something went wrong
-   Allow multi-line log events in combination with format tags
-   Allow repeating format tags
-   Allow format tags in other log levels than `info`

### Fixed

-   Check environment variables for `HOME`, `USER` and `LOCALAPPDATA` first before trying to load configuration files

## [0.2][] – 2018-12-17

Happy holidays 🎄

### Added

-   Print version information with `APP -v` and `APP version`
-   Support on Windows operating systems
-   Add PHP extension `calendar` as requirement
-   Log: Do not colorize "info" events
-   Log: Print the following events as output to STDOUT with `printAsOutput()`
-   Log: Print the following events as messages to STDERR with `printAsMessage()`
-   Log: Print empty line with `printEmptyLine()`
-   Log: Allow method chaining
-   Log: Control text formatting with tags: `<strong>`, `<u>`, `<dim>`, `<fatal>`, `<error>`, `<warning>`, `<notice>`, `<debug>`, `<red>`, `<yellow>`, `<green>`, `<grey>`
-   Make arguments available in configuration settings
-   `IO::in()`: Print message/question to STDERR instead of STDOUT
-   Respect environment variables to disable colored output
-   Also, disable colors if there is no TTY available

### Fixed

-   Let user overwrite a configuration setting if it is an indexed array

## 0.1 – 2018-04-24

Initial release

[Unreleased]: https://github.com/bheisig/cli/compare/0.3...HEAD
[0.3]: https://github.com/bheisig/cli/compare/0.2...0.3
[0.2]: https://github.com/bheisig/cli/compare/0.1...0.2


src/Log.php
<?php

/**
 * Copyright (C) 2018-19 Benjamin Heisig
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Benjamin Heisig <https://benjamin.heisig.name/>
 * @copyright Copyright (C) 2018-19 Benjamin Heisig
 * @license http://www.gnu.org/licenses/agpl-3.0 GNU Affero General Public License (AGPL)
 * @link https://github.com/bheisig/cli
 */

declare(strict_types=1);

namespace bheisig\cli;

use \RuntimeException;

/**
 * Logger
 */
class Log {

    /**
     * Log level: fatal error
     *
     * @var int
     */
    const FATAL = 1;

    /**
     * Log level: error
     *
     * @var int
     */
    const ERROR = 2;

    /**
     * Log level: warning
     *
     * @var int
     */
    const WARNING = 4;

    /**
     * Log level: notice
     *
     * @var int
     */
    const NOTICE = 8;

    /**
     * Log level: info
     *
     * @var int
     */
    const INFO = 16;

    /**
     * Log level: debug
     *
     * @var int
     */
    const DEBUG = 32;

    /**
     * Log level: everything
     *
     * @var int
     */
    const ALL = 63;

    /**
     * Print as output
     *
     * @var string
     */
    const PRINT_AS_OUTPUT = 'STDOUT';

    /**
     * Print as message
     *
     * @var string
     */
    const PRINT_AS_MESSAGE = 'STDERR';

    /**
     * Log Levels
     *
     * @var array
     */
    protected static $levels = [
        self::FATAL => 'fatal',
        self::ERROR => 'error',
        self::WARNING => 'warning',
        self::NOTICE => 'notice',
        self::INFO => 'info',
        self::DEBUG => 'debug'
    ];

    /**
     * Current verbosity
     *
     * @var int
     */
    protected $verbosity;

    /**
     * Colorize output?
     *
     * @var bool Defaults to true
     */
    protected $colorize = true;

    /**
     * Print to standard output or error?
     *
     * @var string Default to standard output
     */
    protected $output = self::PRINT_AS_OUTPUT;

    /**
     * ANSI color codes
     *
     * @var array
     */
    protected $colors = [
        self::FATAL => '0;31',
        self::ERROR => '0;31',
        self::WARNING => '1;33',
        self::NOTICE => '1;33',
        self::DEBUG => '2'
    ];

    /**
     * Constructor
     *
     * @param int $verbosity Verbosity; defaults to everything except debug messages
     * @param bool $colorize Colorize output? Defaults to true
     */
    public function __construct(int $verbosity = self::ALL & ~self::DEBUG, bool $colorize = true) {
        $this->setVerbosity($verbosity);
        $this->printColors($colorize);
    }

    /**
     * Set verbosity
     *
     * @param int $verbosity Verbosity
     *
     * @return self Returns itself
     */
    public function setVerbosity(int $verbosity): self {
        $this->verbosity = $verbosity;
        return $this;
    }

    /**
     * Colorize output?
     *
     * @param bool $colorize Decision
     *
     * @return self Returns itself
     */
    public function printColors(bool $colorize): self {
        $this->colorize = $colorize;
        return $this;
    }

    /**
     * Print the following events as output to STDOUT
     *
     * @return self Returns itself
     */
    public function printAsOutput(): self {
        $this->output = self::PRINT_AS_OUTPUT;
        return $this;
    }

    /**
     * Print the following events as messages to STDERR
     *
     * @return self Returns itself
     */
    public function printAsMessage(): self {
        $this->output = self::PRINT_AS_MESSAGE;
        return $this;
    }

    /**
     * Log event. It provides the same functionality as sprintf() by passing three or more arguments.
     *
     * @param int $level Event level. One of the following class constants:
     * DEBUG, INFO, NOTICE, WARNING, ERROR or FATAL.
     * @param string $value What to be formatted
     * @param mixed ...$args (Optional) One or more arguments
     *
     * @return self Returns itself
     *
     * @see sprintf()
     */
    public function event(int $level, string $value, ...$args): self {
        if ($level & $this->verbosity) {
            $message = $value;

            if (count($args) > 0) {
                $message = call_user_func_array(
                    'sprintf',
                    array_merge([$value], $args)
                );
            }

            if ($this->colorize) {
                switch ($level) {
                    case self::FATAL:
                        $message = "<fatal>$message</fatal>";
                        break;
                    case self::ERROR:
                        $message = "<error>$message</error>";
                        break;
                    case self::WARNING:
                        $message = "<warning>$message</warning>";
                        break;
                    case self::NOTICE:
                        $message = "<notice>$message</notice>";
                        break;
                    case self::DEBUG:
                        $message = "<debug>$message</debug>";
                        break;
                }
            }

            $message = $this->formatText($message);

            $this->flush($message);
        }

        return $this;
    }

    /**
     * Log fatal event. It provides the same functionality as sprintf() by passing two or more arguments.
     *
     * @param string $value What to be formatted
     * @param mixed ...$args (Optional) One or more arguments
     *
     * @return self Returns itself
     *
     * @see sprintf()
     */
    public function fatal(string $value, ...$args): self {
        return call_user_func_array(
            [__CLASS__, 'event'],
            array_merge([self::FATAL, $value], $args)
        );
    }

    /**
     * Log error event. It provides the same functionality as sprintf() by passing two or more arguments.
     *
     * @param string $value What to be formatted
     * @param mixed ...$args (Optional) One or more arguments
     *
     * @return self Returns itself
     *
     * @see sprintf()
     */
    public function error(string $value, ...$args): self {
        return call_user_func_array(
            [__CLASS__, 'event'],
            array_merge([self::ERROR, $value], $args)
        );
    }

    /**
     * Log warning event. It provides the same functionality as sprintf() by passing two or more arguments.
     *
     * @param string $value What to be formatted
     * @param mixed ...$args (Optional) One or more arguments
     *
     * @return self Returns itself
     *
     * @see sprintf()
     */
    public function warning(string $value, ...$args): self {
        return call_user_func_array(
            [__CLASS__, 'event'],
            array_merge([self::WARNING, $value], $args)
        );
    }

    /**
     * Log warning event. It provides the same functionality as sprintf() by passing two or more arguments.
     *
     * @param string $value What to be formatted
     * @param mixed ...$args (Optional) One or more arguments
     *
     * @return self Returns itself
     *
     * @see sprintf()
     */
    public function notice(string $value, ...$args): self {
        return call_user_func_array(
            [__CLASS__, 'event'],
            array_merge([self::NOTICE, $value], $args)
        );
    }

    /**
     * Log info event. It provides the same functionality as sprintf() by passing two or more arguments.
     *
     * @param string $value What to be formatted
     * @param mixed ...$args (Optional) One or more arguments
     *
     * @return self Returns itself
     *
     * @see sprintf()
     */
    public function info(string $value, ...$args): self {
        return call_user_func_array(
            [__CLASS__, 'event'],
            array_merge([self::INFO, $value], $args)
        );
    }

    /**
     * Log debug event. It provides the same functionality as sprintf() by passing two or more arguments.
     *
     * @param string $value What to be formatted
     * @param mixed ...$args (Optional) One or more arguments
     *
     * @return self Returns itself
     *
     * @see sprintf()
     */
    public function debug(string $value, ...$args): self {
        return call_user_func_array(
            [__CLASS__, 'event'],
            array_merge([self::DEBUG, $value], $args)
        );
    }

    /**
     * Print empty line
     *
     * @return self Returns itself
     */
    public function printEmptyLine(): self {
        $this->flush('');

        return $this;
    }

    /**
     * Format text
     *
     * @param string $text Unformatted text
     * @return string Formatted text
     * @throws RuntimeException on error
     */
    protected function formatText(string $text): string {
        $syntax = [
            '/\<strong\>/m' => "\033[1m",
            '/\<u\>/m' => "\033[4m",
            '/\<dim\>/m' => "\033[2m",
            '/\<fatal\>/m' => "\033[" . $this->colors[self::FATAL] . "m",
            '/\<error\>/m' => "\033[" . $this->colors[self::ERROR] . "m",
            '/\<warning\>/m' => "\033[" . $this->colors[self::WARNING] . "m",
            '/\<notice\>/m' => "\033[" . $this->colors[self::NOTICE] . "m",
            '/\<debug\>/m' => "\033[" . $this->colors[self::DEBUG] . "m",
            '/\<red\>/m' => "\033[31m",
            '/\<yellow\>/m' => "\033[33m",
            '/\<green\>/m' => "\033[32m",
            '/\<grey\>/m' => "\033[37m",
            '/\<\/([a-z]+)\>/m' => "\033[0m",
        ];

        $replacements = '';

        if ($this->colorize) {
            $replacements = $syntax;
        }

        $result = preg_replace(
            array_keys($syntax),
            $replacements,
            $text
        );

        if (!is_string($result)) {
            throw new RuntimeException('Unable to format text', ExitApp::RUNTIME_ERROR);
        }

        return $result;
    }

    /**
     * Print message to STDOUT or STDERR
     *
     * @param string $message Message
     */
    protected function flush(string $message) {
        switch ($this->output) {
            case self::PRINT_AS_OUTPUT:
                IO::out($message);
                break;
            case self::PRINT_AS_MESSAGE:
                IO::err($message);
                break;
        }
    }

}
